src/nls/index.ts

```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Internal options.
+ * Will be initialized in `config()`.
+ */
+const options: IInternalOption = Object.create(null);
+
+/**
+ * Updates the in-memory NLS bundle.
+ * @param locales An array of locale IDs. The default locale will be appended.
+ */
+function cacheBundle(locales: VisualStudioCodeLocaleId[] = []): void {
+    if (options.locale) {
+        locales.push(options.locale); // Fallback.
+    }
+
+    // * We always provide `package.nls.json`.
+    // * Reverse the return value, so that we can build a bundle with nice fallback by a simple loop.
+    const files = resolveResource(options.extensionPath, "package.nls", "json", locales)!.reverse() as readonly string[];
+    for (const path of files) {
+        try {
+            Object.assign<typeof resolvedBundle, typeof resolvedBundle>(resolvedBundle, readJsonFile<INlsBundle>(path));
+        } catch (error) {
+            console.error(error); // Log, and ignore the bundle.
+        }
+    }
+}
+
+/**
+ * @param message A composite format string.
+ * @param args An array of objects to format.
+ */
+function format(message: string, ...args: Primitive[]): string {
+    if (args.length === 0) {
+        return message;
+    } else {
+        return message.replace(/\{(0|[1-9]\d*?)\}/g, (match: string, index: string): string => {
+            // `index` is zero-based.
+            return args.length > +index ? String(args[+index]) : match;
+        });
+    }
+}
+
+//#endregion Private
+
+//#region Public
+
+export const localize: IFuncLocalize = function (key: string, ...args: Primitive[]): string {
+    if (options.cacheResolution) {
+        const msg: string | undefined = resolvedBundle[key];
+        return msg === undefined ? "[" + key + "]" : format(msg, ...args);
+    } else {
+        // When in development mode, hot reload, and reveal the key.
+        cacheBundle();
+        const msg: string | undefined = resolvedBundle[key];
+        return msg === undefined ? "[" + key + "]" : "[" + key.substring(key.lastIndexOf(".") + 1) + "] " + format(msg, ...args);
+    }
+};
+
+/**
+ * Configures the NLS module.
+ *
+ * You should only call it **once** in the application entry point.
+ */
+export function config(opts: IConfigOption) {
+    if (opts.locale) {
+        options.locale = opts.locale;
+    } else {
+        try {
+            const vscodeOptions = JSON.parse(process.env.VSCODE_NLS_CONFIG as string) as IVscodeNlsConfig;
+            options.locale = vscodeOptions.locale as any;
+        } catch (error) {
+            // Log, but do nothing else, in case VS Code suddenly changes their mind, or we are not in VS Code.
+            console.error(error);
+        }
+    }
+
+    options.extensionPath = opts.extensionContext.extensionPath;
+    options.cacheResolution = opts.extensionContext.extensionMode !== vscode.ExtensionMode.Development;
+
+    // Load and freeze the cache when not in development mode.
+    if (options.cacheResolution) {
```

Comment:
Having seen all the usages of `cacheResolution`, I guess we can just name it `devMode` (or `useCachedNlsBundle`)?

Reply:
The naming is intentional, though my comments didn't make it clear.

It leaves room for isolating logic, and introducing a `bundleCache` next to `resolvedBundle` in the future.

Reply:
👌